docs/ckstatus.man
.\" Automatically generated from an mdoc input file.  Do not edit.
.\" Copyright (c) 2024-2025 Dieter Baron and Thomas Klausner.
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above
.\"    copyright notice, this list of conditions and the following
.\"    disclaimer in the documentation and/or other materials provided
.\"    with the distribution.
.\" 3. The name of the author may not be used to endorse or promote
.\"    products derived from this software without specific prior
.\"    written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THOMAS KLAUSNER ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE FOUNDATION OR
.\" CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
.\" LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF
.\" USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
.\" ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT
.\" OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.TH "CKSTATUS" "1" "January 9, 2025" "NiH" "General Commands Manual"
.nh
.if n .ad l
.SH "NAME"
\fBckstatus\fR
\- report on
ckmame(1)
runs
.SH "SYNOPSIS"
.HP 9n
\fBckstatus\fR
[\fB\-\-all-missing\fR]
[\fB\-\-all-sets\fR]
[\fB\-\-changes\fR]
[\fB\-\-config\fR\ \fIfile\fR]
[\fB\-\-correct\fR]
[\fB\-\-delete\fR]
[\fB\-\-from\fR\ \fIRUN\fR]
[\fB\-h\fR,\ \fB\-\-help\fR]
[\fB\-\-list-sets\fR]
[\fB\-\-mia-have\fR]
[\fB\-\-missing\fR]
[\fB\-\-run\fR\ \fIRUN\fR]
[\fB\-\-runs\fR]
[\fB\-\-set\fR\ \fIpattern\fR]
[\fB\-\-to\fR\ \fIRUN\fR]
[\fB\-\-status-db\fR\ \fIdbfile\fR]
[\fB\-\-summary\fR]
[\fB\-V\fR,\ \fB\-\-version\fR]
.SH "DESCRIPTION"
\fBckstatus\fR
is a program to report on the results and/or changes of
ckmame(1)
runs on a ROM set.
The data is read from the
\fI.ckmame-status.db\fR
file in the current directory, or the place configured in the config
file, either for the default or the selected set(s).
It compares the output of different runs, i.e. starts of
ckmame(1).
.SS "OPTIONS"
Supported options:
.TP 22n
\fB\-\-all-missing\fR
List all missing ROMs.
.TP 22n
\fB\-\-all-sets\fR
Report on all configured sets.
.TP 22n
\fB\-\-changes\fR
List the changes in between two runs.
By default, the newest two runs are compared, but the runs to compare
can be chosen with \fB\-\-from\fR \fIRUN\fR
and \fB\-\-to\fR \fIRUN\fR.
.TP 22n
\fB\-\-config\fR \fIfile\fR
Read the configuration from
\fIfile\fR.
.TP 22n
\fB\-\-correct\fR
List all correct ROMs.
.TP 22n
\fB\-h\fR, \fB\-\-help\fR
Show a short help.
.TP 22n
\fB\-\-delete\fR
Delete the run specified by the \fB\-\-run\fR \fIRUN\fR
option (required to avoid accidental data loss).
.TP 22n
\fB\-\-from\fR \fIRUN\fR
Select the base run for the comparison with \fB\-\-changes\fR.
.TP 22n
\fB\-\-list-sets\fR
List all configured sets, for the \fB\-\-set\fR
option.
.TP 22n
\fB\-\-mia-have\fR
List all games containing ROMs that are MIA according to the database.
.TP 22n
\fB\-\-missing\fR
List all missing ROMs except those that are MIA according to the
database.
.TP 22n
\fB\-\-run\fR \fIRUN\fR
Select the run on which to report, or which to delete
(for \fB\-\-delete\fR).
.TP 22n
\fB\-\-runs\fR
List all available runs.
.TP 22n
\fB\-\-set\fR \fIpattern\fR
Report on all sets matching
\fIpattern\fR.
.TP 22n
\fB\-\-status-db\fR \fIdbfile\fR
Use
\fIdbfile\fR
instead of the default
\fI.ckmame-status.db\fR.
.TP 22n
\fB\-\-summary\fR
Print summary information on the number of missing games.
.TP 22n
\fB\-\-to\fR \fIRUN\fR
Select the target run for the comparison with \fB\-\-changes\fR.
.TP 22n
\fB\-V\fR, \fB\-\-version\fR
Report the version of
\fBckstatus\fR.
.SH "FILES"
By default, the run information is read from
\fI.ckmame-status.db\fR.
.SH "SEE ALSO"
ckmame(1),
ckmamerc(5)
.SH "AUTHORS"
\fBckstatus\fR
was written by
Dieter Baron <\fIdillo@nih.at\fR>
and
Thomas Klausner <\fIwiz@gatalith.at\fR>.


NEWS
0.12 (2015-05-03)

- add support for on-disk caches to speed up consecutive runs
- support for ROMs in directories instead of in zip archives
- fix all fixable ROM errors automatically in one run
- speed up database operations by preparing statements only once
- remove torrentzip support (removed in libzip)

0.11 (2011-02-20)

- MESS Software Lists support
- retire mkmamedb-xmame.sh, use "mame -listxml | mkmamedb" instead
- link/copy disks from extra if -j (--delete-found) isn't specified
- ckmame: display current game on SIGINFO (Ctrl-T)
- removed ROMPATH support since mame uses a path in the config file nowadays

0.10 (2008-07-25)

- use SQLite3 instead of Berkeley DB
- optionally TorrentZip ROM set
- mkmamedb: add support for ROM Management Datafile format
- mkmamedb: add support to read dat files from zip archives
- mkmamedb: extract version and game description from listxml output
- ckmame: fix finding ROMs that need detectors from superfluous and extra
- change to 3-clause BSD license

0.9 (2007-06-04)

- add support for CMPro XML header skip detectors
- fix handling of zero byte ROMs
- dumpgame: brief option: no ROM/disk info
- mkmamedb: add support for reading Romcenter dat files
- mkmamedb: warn about multiple games with same name
- mkmamedb: create CMPro dat files
- mkmamedb: create mame db or CMPro dat file from zip archives
- ckmame: add option to keep files present in old ROM database

0.8 (2006-05-18)

- clean up superfluous files and extra directories
- use additional DB listing files that exist elsewhere
- read games to check from file
- create DB from multiple dat files
- omit games matching shell glob patterns from database
- don't open archives or disk images multiple times
- don't accept disk images that need a parent
- move unknown files to directory `unknown', not to `garbage' in ROM path

0.7 (2005-12-28)

- fix all fixable errors in at most two runs
- rename disks
- support searching for files in additional directories
- detect faked ROMs (correct CRC, wrong MD5/SHA1)
- adapt mkmamedb-xmame.sh for MAME version numbers of 0.100 and above
- dropped support for inferring flags/status from CRC (old MAME versions)
- deprecated -u/-U command line options
- simplify ROM matching logic
- many new regression tests, causing
- various bug fixes

0.6 (2005-06-17)

- libzip is distributed separately
- Support MAME/MESS -listxml output
- Adapt for MAME -listinfo output changes (honor flags baddump and nodump).
- Check disk images (chd files).
- Superfluous merged into ckmame. ckmame now by default reports
  extra files.  Also works for samples.
- add file format version to database and don't use incompatible database
- dumpgame: find game by checksum of one of its ROMs
- add script to create database by running xmame

0.5 (2003-01-30)

MESS, Impact, Raine, and others based on the CMPro format now supported.
Man page added.
Bug fixes.

0.4.1 (1999-10-06)

If a a clone zip didn't exist, it was not created.
The DOS port now really can fix romsets.

0.4 (1999-09-23)

The whole program.


src/ckstatus.cc
/*
  ckstatus.cc -- print info about last runs (from data base)
  Copyright (C) 1999-2018 Dieter Baron and Thomas Klausner

  This file is part of ckmame, a program to check rom sets for MAME.
  The authors can be contacted at <ckmame@nih.at>

  Redistribution and use in source and binary forms, with or without
  modification, are permitted provided that the following conditions
  are met:
  1. Redistributions of source code must retain the above copyright
     notice, this list of conditions and the following disclaimer.
  2. Redistributions in binary form must reproduce the above copyright
     notice, this list of conditions and the following disclaimer in
     the documentation and/or other materials provided with the
     distribution.
  3. The name of the author may not be used to endorse or promote
     products derived from this software without specific prior
     written permission.

  THIS SOFTWARE IS PROVIDED BY THE AUTHORS ``AS IS'' AND ANY EXPRESS
  OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
  WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
  ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY
  DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
  DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
  GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
  IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
  OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN
  IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/

#include "CkStatus.h"

#include <algorithm>
#include <cerrno>
#include <cstring>
#include <string>

#include "compat.h"

#include "ArchiveDir.h"
#include "Commandline.h"
#include "Exception.h"
#include "RomDB.h"
#include "Stats.h"
#include "globals.h"
#include "util.h"

std::vector<Commandline::Option> ckstatus_options = {
    Commandline::Option("all-missing", "list missing games"),
    Commandline::Option("changes", "list changes between runs"),
    Commandline::Option("correct", "list correct games"),
    Commandline::Option("delete", "delete specific run"),
    Commandline::Option("from", "RUN", "first run to compare with --changes"),
    Commandline::Option("mia-have", "list games containing files marked MIA"),
    Commandline::Option("missing", "list missing games not marked MIA"),
    Commandline::Option("run", "RUN", "use information from run RUN (default: latest)"),
    Commandline::Option("runs", "list runs"),
    Commandline::Option("summary", "print summary"),
    Commandline::Option("to", "RUN", "second run to compare with --changes")};


std::unordered_set<std::string> ckstatus_used_variables = {"status_db", "report_status"};

int main(int argc, char** argv) {
    auto command = CkStatus();

    return command.run(argc, argv);
}


CkStatus::CkStatus() : Command("ckstatus", "", ckstatus_options, ckstatus_used_variables) {}

void CkStatus::global_setup(const ParsedCommandline& commandline) {
    for (const auto& option : commandline.options) {
        if (option.name == "all-missing") {
            specials.insert(ALL_MISSING);
        }
        else if (option.name == "changes") {
            specials.insert(CHANGES);
        }
        else if (option.name == "correct") {
            specials.insert(CORRECT);
        }
        else if (option.name == "delete") {
            specials.insert(DELETE_RUN);
        }
        else if (option.name == "from") {
            run_from = std::stoll(option.argument);
        }
        else if (option.name == "mia-have") {
            specials.insert(LIST_MIA);
        }
        else if (option.name == "missing") {
            specials.insert(MISSING);
        }
        else if (option.name == "run") {
            run_id = std::stoll(option.argument);
        }
        else if (option.name == "runs") {
            specials.insert(RUNS);
        }
        else if (option.name == "summary") {
            specials.insert(SUMMARY);
        }
        else if (option.name == "to") {
            run_to = std::stoll(option.argument);
        }
    }

    if (specials.empty()) {
        specials.insert(SUMMARY);
    }
}

bool CkStatus::execute(const std::vector<std::string>& arguments) {
    if (configuration.status_db == "none") {
        throw Exception("No status database configured.");
    }

    if (configuration.report_status == false) {
        /* this set should be ignored according by the config */
        return true;
    }

    try {
        status_db = std::make_shared<StatusDB>(configuration.status_db, DBH_READ);
    }
    catch (const std::exception& e) {
        throw Exception("Error opening status database: " + std::string(e.what()));
    }

    for (auto key : specials) {
        switch (key) {
        case ALL_MISSING:
            list_games({GS_MISSING, GS_MISSING_BEST, GS_PARTIAL, GS_PARTIAL_BEST, GS_PARTIAL_MIA, GS_PARTIAL_BEST_MIA});
            break;

        case CHANGES:
            print_changes();
            break;

        case CORRECT:
            list_games({GS_CORRECT, GS_CORRECT_MIA});
            break;

        case DELETE_RUN:
            delete_run();

        case LIST_MIA:
            list_games({GS_CORRECT_MIA, GS_PARTIAL_MIA});
            break;

        case MISSING:
            list_games({GS_MISSING, GS_PARTIAL, GS_PARTIAL_MIA});
            break;

        case RUNS:
            list_runs();
            break;

        case SUMMARY:
            list_summary();
        }
    }
    return true;
}

bool CkStatus::cleanup() {
    status_db = nullptr;

    return true;
}

int CkStatus::get_run_id() {
    if (run_id) {
        return *run_id;
    }

    auto actual_run_id = status_db->latest_run_id(false);
    if (!actual_run_id) {
        throw Exception("No runs.");
    }
    return *actual_run_id;
}

int CkStatus::get_run_to() {
    if (run_to) {
        return *run_to;
    }

    auto actual_run_to = status_db->latest_run_id(false);
    if (!actual_run_to) {
        throw Exception("No runs.");
    }
    return *actual_run_to;
}


int CkStatus::get_run_from() {
    if (run_from) {
        return *run_from;
    }

    auto actual_run_from = status_db->latest_run_id(true);
    if (!actual_run_from) {
        throw Exception("Only one run.");
    }
    return *actual_run_from;
}


void CkStatus::delete_run() {
    if (!run_id) {
        throw Exception("--run option must be specified for --delete");
    }

    status_db->delete_run(*run_id);
}


void CkStatus::list_games(const std::unordered_set<GameStatus>& status) {
    auto games = status_db->get_games_by_status(get_run_id(), status);
    sort_strings_case_insensitive(games);
    for (const auto& game : games) {
        output.message(game);
    }
}


void CkStatus::list_runs() {
    auto runs = status_db->list_runs();

    for (const auto& run : runs) {
        char date_string[20];
        strftime(date_string, sizeof(date_string), "%Y-%m-%d %H:%M:%S", localtime(&run.date));
        output.message("%" PRId64 ": %s", run.run_id, date_string);
    }
}

void CkStatus::list_summary() {
    auto counts = status_db->get_run_status_counts(get_run_id());

    uint64_t total = 0;
    uint64_t missing = 0;
    uint64_t can_improve = 0;

    for (const auto [status, count] : counts) {
        total += count;

        if (GS_HAS_MISSING(status)) {
            missing += count;
        }

        if (GS_CAN_IMPROVE(status)) {
            can_improve += count;
        }
    }

    // best = total - missing_mia
    // best = missing + can_improve

    auto status = std::string{};
    auto best = std::string{};
    if (missing - can_improve > 0) {
        best = " (best: " + std::to_string(missing - can_improve) + ")";
    }

    if (missing == 0) {
        status = "complete";
    }
    else {
        status = std::to_string(missing) + " / " + std::to_string(total) + " missing";
    }
    output.message(status + best);
}


void CkStatus::print_changes() {
    auto diffs = RunDiff(get_run_from(), get_run_to());

    diffs.compute();
}

void CkStatus::RunDiff::compute() {
    insert_run(false);
    insert_run(true);

    std::sort(games.begin(), games.end());

    for (const auto& game : games) {
        auto diff = diffs.at(game.checksum);
        if (diff.new_info && is_complete(diff.new_info->status)) {
            if (is_complete(diff.new_info->status)) {
                if (!diff.old_info || !is_complete(diff.old_info->status)) {
                    output.message("+ %s", game.name.c_str());
                }
            }
        }
        else if (diff.old_info && is_complete(diff.old_info->status)) {
            output.message("- %s", game.name.c_str());
        }
    }
}


void CkStatus::RunDiff::insert_run(bool old) {
    auto infos = status_db->get_games(old ? run_from : run_to);

    for (const auto& info : infos) {
        auto it = diffs.find(info.checksum);
        if (it != diffs.end()) {
            if (old) {
                it->second.old_info = info;
            }
            else {
                it->second.new_info = info;
            }
        }
        else {
            games.push_back({info.name, info.checksum});
            auto diff = GameDiff{};
            if (old) {
                diff.old_info = info;
            }
            else {
                diff.new_info = info;
            }
            diffs[info.checksum] = diff;
        }
    }
}
